Refactor: Make backend's outbuffer a little bit safer

Geod24 opened the pull request:
Better reviewed commit-by-commit.

The first commit is the largest one, but it's also a mechanical change from "size" to "length" to bring the naming in line with the D convention, where "length" is the used length, while "capacity" is the total size of the buffer (not exposed here).

This is a range of commit made while working on various things in the backend (DWARF, ABI issues), hence the disconnect for some commit. However each commit should be self-explanatory (e.g. the DWARF documentation one).

Geod24 committed:
dmd.backend.Outbuffer: Rename size to length

Follow the naming convention of dmd.root.outbuffer : OutBuffer

Geod24 committed:
backend.outbuffer: Expose opSlice for safe usage

So far lots of calling code has been doing magic with Outbuffer,
accessing the internal pointers and/or iterating over them.
Exposing opSlice is the first step in providing a safer interface.

Geod24 committed:
backend.outbuffer: Use the slice overload to copy other outbuffer

Geod24 committed:
backend.dwarf: Use slices for strings

Geod24 committed:
Trivial: Fix backend.dwarf module documentation

Geod24 committed:
backend.outbuffer: Eliminate clearn method

It doesn't (and cannot) expose a safe interface. It was only used in 3 places where writezeros can be used instead.

Geod24 committed:
backend.code3: Make MiniCodeBuf private

Geod24 committed:
backend.outbuffer: Remove unused or dangerous method

flush is a no-op, writeChar is quite misleading in its name,
write(const ubyte*) just assume its argument is a NUL-terminated string,
and the unused toString has no place in a buffer used for binary data.

thewilsonator commented on src/dmd/backend/mscoffobj.d:
```diff
@@ -2077,16 +2077,12 @@ static if (0)
             //seg,offset,nbytes,p);
-    buf.setsize(cast(uint)offset);
-    buf.reserve(nbytes);
+    buf.position(cast(size_t)offset, nbytes);
```
Not that It particularly matters or this PR, but this whole function `MsCoffObj_bytes ` appears to be an exact duplicate of `Obj_bytes` in both `machobj.d` and `elfobj.d`

Geod24 replied:
Yep! The backend has a lot of duplications like this... I'm trying to get rid of *some* of it to make my issues easier to fix.

thewilsonator commented:
There seems to be a lot of errors that looks like 
```
dmd/backend/elfobj.d(547): Error: function `dmd.backend.global.size(uint)` is not callable using argument types `(Outbuffer*)`
dmd/backend/elfobj.d(547):        cannot pass argument `section_names` of type `Outbuffer*` to parameter `uint`
```

thewilsonator reviewed (approved):
Other than the failing to compile this looks fine to me.

Geod24 commented:
> There seems to be a lot of errors that looks like

Yep, I could only compile on Mac. Should be green now.

There were a couple of changes:
- Reworked `outbuf` destructor in the second commit;
- Reverted the change from `writeString` to `write`, instead introduced 2 `writeString` overloads that write a NUL-terminator (affected the DWARF switch to slices);
- A couple more place where `size` => `length` was missing

benjones commented:
Should the slicing operations with strlen use root.string.toDString instead?

Geod24 commented:
> Should the slicing operations with strlen use root.string.toDString instead?

Ideally yes. However I had some issues when adding `root` import in the backend because `betterC` (e.g. importing array didn't work) so decided to keep it simple for now.
